File: examples/test-script-walkers.jl
# Description: 
# - Performance test script example with N random walkers.
#   - Great to see how performance scales with number of objects.
# Usage: `julia --project examples/test-script-walkers.jl --nwalkers 200`
# Expected: Should run without errors and print fps to console.
# Output Example:
#   N_walkers: 100
#   fps: 41.94161430139263
#   fps: 42.10860020065973
#   fps: 42.122480158458885
#   Exiting after 240 frames.


# examples/random-walkers-1.jl
using Pkg
Pkg.activate(joinpath(@__DIR__, ".."))
import SDLProcessing
const P = SDLProcessing
import SimpleDirectMediaLayer
const SDL2 = SimpleDirectMediaLayer.LibSDL2

## --- Get N from ARGS ---
# Usage: `julia --project examples/test-script-walkers.jl --nwalkers 200`
global N_walkers = 10000
for (i, arg) in enumerate(ARGS)
    if arg == "--nwalkers" && i < length(ARGS)
        global N_walkers = parse(Int, ARGS[i+1])
    end
end

## --- Walker Object ---
mutable struct Walker
    x::Float64
    y::Float64
end

# Create a global list of walkers
walkers = Walker[]

## --- Sketch Definition ---

P.onsetup() do
    global N_walkers
    println("N_walkers: ", N_walkers)

    P.create_window(800, 600; title="Random Walkers")

    # Initialize N_walkers walkers at random positions
    for _ in 1:N_walkers
        x = rand() * P.SKETCH.width
        y = rand() * P.SKETCH.height
        push!(walkers, Walker(x, y))
    end
end

nprints = 0

P.ondraw() do
    
    global nprints

    P.background(P.color(0, 0, 0)) # Black background

    # --- Update and Draw Walkers ---
    P.fill(P.color(128, 128, 128)) # Gray color for walkers
    P.noStroke()

    for walker in walkers
        # Move walker
        walker.x += (rand() - 0.5) * 2
        walker.y += (rand() - 0.5) * 2

        # Periodic boundaries
        if walker.x > P.SKETCH.width
            walker.x = 0
        elseif walker.x < 0
            walker.x = P.SKETCH.width
        end

        if walker.y > P.SKETCH.height
            walker.y = 0
        elseif walker.y < 0
            walker.y = P.SKETCH.height
        end

        # Draw walker as a pixel
        P.rect(walker.x, walker.y, 1, 1)
    end

    # do nothing
    if P.isFrameRateDataReady() && mod(P.SKETCH.frameCount, 60) == 0
        println("fps: ", P.frameRate())
        nprints += 1
    end

    # quit after some prints
    if nprints > 3
        P.SKETCH._should_quit = true
        println("Exiting after $(P.SKETCH.frameCount) frames.")
    end
end

## --- Run the Sketch ---
P.run_sketch()

File: test/scripts/js/test-script-walkers.js
// Description: 
// - Performance test script example with N random walkers.
// - Great to see how performance scales with number of objects.
// Based on: test/scripts/jl/test-script-walkers.jl

// In browser, we can't easily pass command line args.
// Change this value to test with different numbers of walkers.
const N_walkers = 10000;

let walkers = [];
let nprints = 0;
const maxPrints = 3;

class Walker {
  constructor() {
    this.x = random(width);
    this.y = random(height);
  }

  step() {
    this.x += random(-1, 1);
    this.y += random(-1, 1);

    // Periodic boundaries
    if (this.x > width) this.x = 0;
    if (this.x < 0) this.x = width;
    if (this.y > height) this.y = 0;
    if (this.y < 0) this.y = height;
  }

  display() {
    rect(this.x, this.y, 1, 1);
  }
}

function setup() {
  console.log(`N_walkers: ${N_walkers}`);
  createCanvas(800, 600);

  for (let i = 0; i < N_walkers; i++) {
    walkers.push(new Walker());
  }
}

function draw() {
  background(0); // Black background

  fill(128); // Gray color for walkers
  noStroke();

  for (let walker of walkers) {
    walker.step();
    walker.display();
  }

  // p5.js's frameRate() is available but more accurate after some frames.
  if (frameCount > 10 && frameCount % 60 === 0) {
    console.log(`fps: ${frameRate()}`);
    nprints++;
  }

  if (nprints >= maxPrints) {
    console.log(`Exiting after ${frameCount} frames.`);
    noLoop();
  }
}
